### DISCUSSION TITLE
How to limit or set size for the default ThreadPoolExecutor which be used to run blocking background task in FastAPI?

### DISCUSSION THREAD

**Discussion by Nohysiwe** [fastapi/fastapi#6728] [Questions] [answered]
I noticed that blocking background task was running in the default ThreadPoolExecutor for the current thread's asyncio.Eventloop, so how to limit or set size of the default ThreadPoolExecutor. ：)

**Comment by insomnes** [comment]
@Nohysiwe 
You can try to set default ThreadPoolExecutor in your main file

```python3
from concurrent.futures import ThreadPoolExecutor
import asyncio

loop = asyncio.get_running_loop()
loop.set_default_executor(ThreadPoolExecutor(max_workers=4))
```

**Comment by Nohysiwe** [comment]
@insomnes 
Thank you.

**Comment by Nohysiwe** [comment]
@insomnes
I think that using `loop = asyncio.get_event_loop()` is better than `loop = asyncio.get_running_loop()` in here.

**Comment by Nohysiwe** [comment]
And I find that I can use `ProcessPoolExecutor` same as `ThreadPoolExecutor` . But it's not a good idea when the app run with multi `worker`.

**Comment by Nohysiwe** [comment]
@insomnes
I find that it's ineffective to set default ThreadPoolExecutor in main file. Because the loop used by uvicorn is from `asyncio.new_event_loop()`.  It run after import my main file in the code of uvicorn.
``` python
# in uvicorn/loops/asyncio.py/asyncio_setup()
def asyncio_setup():
    if (
        sys.version_info.major >= 3
        and sys.version_info.minor >= 8
        and platform.system() == "Windows"
    ):
        selector = selectors.SelectSelector()
        loop = asyncio.SelectorEventLoop(selector)
        asyncio.set_event_loop(loop)
    else:
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)

```

**Comment by Kludex** [comment]
I'm not sure why are you doing this, but...

You can add the @insomnes logic inside of this: https://fastapi.tiangolo.com/advanced/events/?h=startup#startup-event

**Comment by Nohysiwe** [comment]
@Kludex 
Thank you for your help :)

**Comment by tiangolo** [comment]
Thanks for the help here @insomnes and @Kludex ! 👏  🙇 

Thanks for reporting back and closing the issue @Nohysiwe  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by Kludex** [comment, marked as answer]
The answer for this question is a bit different at the moment, see https://github.com/encode/starlette/issues/1724#issuecomment-1179063924 for the answer.

**Reply by HenningDinero** [reply]
Just to be clear; that piece of code should be oin the `@app.startup` middleware of a FastAPI application, right (if one wants to limit the number of threads)?
